source/driver-models/Serial.cpp

```diff
@@ -467,8 +469,26 @@ void Serial::printf(const char* format, ...)
                 break;
 
             case 'x':
-            case 'p':
+              lowerCase = true;
             case 'X':
+              for (uint8_t i = 8; i > 0; i--)
+              {
+                  uint8_t temp = ((uint8_t) (val >> ((i - 1) * 4)) & 0x0f) + (uint8_t) '0';
+                  if (temp > '9')
+                  {
+                      if (lowerCase == true)
+                          temp += 39;
+                      else
+                          temp += 7;
+                  }
+                  if (firstDigit == false || temp != '0')
```

Comment:
Thanks for the contribution!

Could you explain the purpose of `firstDigit` to me? Not sure I follow the purpose of the variable.

Reply:
Hi, James,
For example, if the printing value is 0x000ABC12, I wished it prints "ABC12" rather than 000ABC12.
"firstDigit" flag is used to mark if the first non-zero digit is found. To be more precise, its name should be "firstDigitNotFound".
Now I think it is better to change its name to "firstDigitFound", and invert its logic.

Reply:
Ok that makes sense. What if the value is 0?

Reply:
That IS a problem..
I should fix it.

Reply:
I fixed the problem and corrected indent(2 spaces -> 4 spaces). But there was some mistake during my commit operations and here appears 8 more commits(from [Add Pin] to [PR feedback])  which are not mine. I only modified Serial.cpp.

Reply:
@irarykim I think this should just boil down to the following:

```cpp
case 'x':
    lowerCase = true;
case 'X':
    for (uint8_t i = 8; i > 0; i--)
    {
        uint8_t digit = ((uint8_t) (val >> ((i - 1) * 4)) & 0x0f) + (uint8_t) '0';
        if (digit > '9')
        {
            if (lowerCase)
                digit += 39;
            else
                digit += 7;
        }
        if (digit != '0')
        {
            putc((char)digit);
            firstDigitFound = true;
        }
    }
    if (!firstDigitFound)
        putc('0');
    break;
```

It's a bit cheeky of me to ask, but would you be able to test the above works?

Reply:
@jamesadevine  No problem for testing. 
I have just tested and I found it doesn't print '0'. For example, it prints 0xA0B0 as AB.

How about following (reduced) code? It prints all following numbers correctly.
0, 0xABC00, 0x1A02B3, 0x1234ABCD, 0xFFFFFFFF, 0x1234567890(prints 34567890)
```c
    case 'x':
        lowerCase = true;
    case 'X':
        for (uint8_t i = 8; i > 0; i--)
        {
            uint8_t digit = ((uint8_t) (val >> ((i - 1) * 4)) & 0x0f) + (uint8_t) '0';
            if (digit > '9')
            {
                if (lowerCase)
                    digit += 39;
                else
                    digit += 7;
            }
            if (digit != '0')
            {
                putc((char)digit);
                firstDigitFound = true;
            }
            else if (firstDigitFound || i == 1)
                putc((char)digit);
        }
        break;
```
Besides, these code throws following warning during build, because there is no break; before case 'X'.
```
../libraries/codal-core/source/driver-models/Serial.cpp:472:27: warning: this statement may fall through [-Wimplicit-fallthrough=]
                 lowerCase = true;
                 ~~~~~~~~~~^~~~~~
```
In order to fix it, I modified that part and then it doesn't generate that warning.
```
            case 'x':
            case 'X':
                if (*(end-1) == 'x')
                    lowerCase = true;
```

Reply:
Cool! Let's go with your latest version, then.

>../libraries/codal-core/source/driver-models/Serial.cpp:472:27: warning: this statement may fall through [-Wimplicit-fallthrough=]
                 lowerCase = true;
You can add:
```
// fall through
```
above the case statement to get rid of the warning.

Reply:
Thanks for your tip! 
And for being clearer, the "your latest version" means my (reduced) version above, right?

Reply:
Yes, if the reduced version above behaves in the same way, I'd prefer that one! (just me being picky :wink: ) 

Thanks so much!

Reply:
I committed it. 
Never mind, picky maketh programmer~ :)

Reply:
Hahaha, thank you for understanding! :smile:

The git commits look a bit messed up...it seems like some commits have been recommitted with a different identifier. Would you be able to either force push to your branch with just your changes (bit risky) or apply your changes to a new branch forked from master and recreate the PR (pretty safe)?

Reply:
THEN I'LL MERGE IT I PROMISE!!

Reply:
Yes, using a new branch and new PR is also what I thought. That is safe. 
Then I will close this PR and make a new PR, right? Actually it is my first PR. Thanks for your understanding.

Reply:
I closed this PR and made a new PR.